[macOS] Increase contrast setting requires Safari restart for colors to update, system accent color change does not trigger repaint

lilyspiniolas opened the pull request:
#### ebef4b11b989729196e56220ab9fc55580353b3b
<pre>
[macOS] Increase contrast setting requires Safari restart for colors to update, system accent color change does not trigger repaint
<a href="https://bugs.webkit.org/show_bug.cgi?id=297837">https://bugs.webkit.org/show_bug.cgi?id=297837</a>
<a href="https://rdar.apple.com/137778191">rdar://137778191</a>

Reviewed by Aditya Keerthi.

Since CFPrefs direct mode enablement (<a href="https://trac.webkit.org/changeset/258064/webkit)">https://trac.webkit.org/changeset/258064/webkit)</a>,
the web process never recieved notifications for updates to the increase contrast setting
value. As a result, the system color cache was never cleared. To fix this, forward
the notification from the UI process to the web process.

When the system accent color changes on macOS, the color cache is properly cleared,
but no repaint occurs. As a result, the updated color will not be reflected on the
page until a repaint is triggered by other means. To fix this, call new method
`Page::updateControlTintsForAllPages` when the system accent color changes,
which results in the repaint of all controls with control tint.

* LayoutTests/accessibility/mac/increase-contrast-system-color-changes-expected-mismatch.html: Added.
* LayoutTests/accessibility/mac/increase-contrast-system-color-changes.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateControlTintsForAllPages):
(WebCore::Page::updateControlTints):
* Source/WebCore/page/Page.h:
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
(-[WebCoreRenderThemeNotificationObserver init]):
(-[WebCoreRenderThemeNotificationObserver accessibilityDisplayOptionsDidChange:]):
(-[WebCoreRenderThemeNotificationObserver systemColorsDidChange:]):
* Source/WebKit/Shared/AuxiliaryProcess.h:
* Source/WebKit/Shared/Cocoa/AuxiliaryProcessCocoa.mm:
(WebKit::AuxiliaryProcess::increaseContrastPreferenceKey):
(WebKit::increaseContrastPreferenceKey): Deleted.
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::dispatchSimulatedNotificationsForPreferenceChange):

Canonical link: <a href="https://commits.webkit.org/299445@main">https://commits.webkit.org/299445@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1116f4fc0b93627f38ec94cd20cb68891302230f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118976 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38657 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29308 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125177 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71039 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/0a53531c-85ac-4faf-a32f-e94d411a3dc5) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120854 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39353 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47239 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90290 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59799 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4ec352b3-3841-4d89-90a5-b5aa6184c012) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121929 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31376 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106670 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70798 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9569b79c-a72e-41dd-844d-a18091dcfdd3) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30439 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24784 "Found 1 new test failure: inspector/worker/dom-debugger-event-after-terminate-crash.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68829 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100822 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24972 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128216 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45883 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34668 "Found 1 new test failure: imported/w3c/web-platform-tests/custom-elements/pseudo-class-defined-customized-builtins.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98957 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46250 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102890 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98737 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44195 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22195 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42452 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18948 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45753 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51431 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45219 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48565 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46905 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/page/Page.h:
```diff
@@ -374,6 +374,7 @@ class Page : public RefCountedAndCanMakeWeakPtr<Page>, public Supplementable<Pag
     WEBCORE_EXPORT static Ref<Page> create(PageConfiguration&&);
     WEBCORE_EXPORT ~Page();
 
+    WEBCORE_EXPORT static void updateControlTintsForAllPages();
```
Why `WEBCORE_EXPORT`? Don't see any use outside WebCore.

pxlcoder commented on Source/WebCore/page/Page.h:
```diff
@@ -1446,6 +1447,8 @@ class Page : public RefCountedAndCanMakeWeakPtr<Page>, public Supplementable<Pag
 
     bool hasLocalMainFrame();
 
+    WEBCORE_EXPORT void updateControlTints();
```
Same question.

pxlcoder commented on Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
```diff
@@ -237,6 +278,11 @@ @implementation WebCoreRenderThemeBundle
 constexpr int kVisibleBackgroundImageWidth = 1;
 constexpr int kMultipleThumbnailShrinkSize = 2;
 
+RenderThemeCocoa::RenderThemeCocoa()
```
Nit: I'd move this under the static method below, keeping it next to the other methods on `RenderThemeCocoa`.

pxlcoder commented on Source/WebCore/rendering/ios/RenderThemeIOS.mm:
```diff
@@ -1211,6 +1211,12 @@ static void adjustInputElementButtonStyle(RenderStyle& style, const HTMLInputEle
 
+void RenderThemeIOS::platformColorsDidChange()
+{
+    globalCSSValueToSystemColorMap().clear();
```
How does the map get re-populated?

lilyspiniolas replied:
It doesn't, I'll add logic to repopulate it from the UI process.

pxlcoder commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -1383,6 +1383,16 @@ static float currentBacklightLevel()
+        [notificationCenter postNotificationName:NSWorkspaceAccessibilityDisplayOptionsDidChangeNotification object:nil];
+#else
+        RetainPtr notificationCenter = [NSNotificationCenter defaultCenter];
+        [notificationCenter postNotificationName:@"UIAccessibilityDarkerSystemColorsStatusWillChangeNotification" object:nil];
```
`UIAccessibilityDarkerSystemColorsStatusWillChangeNotification` doesn't seem to be a real thing?

pxlcoder commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -1383,6 +1383,16 @@ static float currentBacklightLevel()
+    } else if (key == increaseContrastPreferenceKey()) {
+#if PLATFORM(MAC)
+        RetainPtr notificationCenter = [[NSWorkspace sharedWorkspace] notificationCenter];
+        [notificationCenter postNotificationName:@"NSWorkspaceAccessibilityDisplayOptionsWillChangeNotification" object:nil];
```
Ditto.

pxlcoder commented on Source/WebCore/rendering/mac/RenderThemeMac.mm:
```diff
@@ -99,19 +99,28 @@ - (id)init
     return self;
 }
 
+- (void)accessibilityDisplayOptionsDidChange:(NSNotification *)notification
```
Just to confirm, is this actually necessary? I'm slightly surprised that `NSSystemColorsDidChangeNotification` wouldn't get posted when accessibility display options change.

lilyspiniolas replied:
`NSSystemColorsDidChangeNotification` is posted manually from `WebProcess::dispatchSimulatedNotificationsForPreferenceChange` for accent-color changes only. We could call `systemColorsDidChange` for both accent-color changes and for changes to the increase contrast setting, but in the latter case we would be performing unnecessary additional painting.

pxlcoder commented on LayoutTests/accessibility/mac/increase-contrast-system-color-changes-expected-mismatch.html:
```diff
@@ -0,0 +1,33 @@
+
+            setTimeout(function () {
+                testRunner.notifyDone();
+            }, 50);
```
Can we use `0` or `requestAnimationFrame`? `50` makes it seems like this could be non-deterministic, but `injectAccessibilityPreference` seems to directly call into the preference update method.

Even a nested `requestAnimationFrame` would be preferable to 50, if it works.

lilyspiniolas replied:
This sadly does seem to be non-deterministic. The test fails most of the time when using `0` and fails more than 25% of the time when using `requestAnimationFrame`. Nested `requestAnimationFrame` strangely did not help, either.

lilyspiniolas committed:
[macOS] Increase contrast setting requires Safari restart for colors to update, system accent color change does not trigger repaint
https://bugs.webkit.org/show_bug.cgi?id=297837
rdar://137778191

Reviewed by Aditya Keerthi.

Since CFPrefs direct mode enablement (https://trac.webkit.org/changeset/258064/webkit),
the web process never recieved notifications for updates to the increase contrast setting
value. As a result, the system color cache was never cleared. To fix this, forward
the notification from the UI process to the web process.

When the system accent color changes on macOS, the color cache is properly cleared,
but no repaint occurs. As a result, the updated color will not be reflected on the
page until a repaint is triggered by other means. To fix this, call new method
`Page::updateControlTintsForAllPages` when the system accent color changes,
which results in the repaint of all controls with control tint.

* LayoutTests/accessibility/mac/increase-contrast-system-color-changes-expected-mismatch.html: Added.
* LayoutTests/accessibility/mac/increase-contrast-system-color-changes.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateControlTintsForAllPages):
(WebCore::Page::updateControlTints):
* Source/WebCore/page/Page.h:
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
(-[WebCoreRenderThemeNotificationObserver init]):
(-[WebCoreRenderThemeNotificationObserver accessibilityDisplayOptionsDidChange:]):
(-[WebCoreRenderThemeNotificationObserver systemColorsDidChange:]):
* Source/WebKit/Shared/AuxiliaryProcess.h:
* Source/WebKit/Shared/Cocoa/AuxiliaryProcessCocoa.mm:
(WebKit::AuxiliaryProcess::increaseContrastPreferenceKey):
(WebKit::increaseContrastPreferenceKey): Deleted.
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::dispatchSimulatedNotificationsForPreferenceChange):

Canonical link: https://commits.webkit.org/299445@main
